lib/chat_web/live/main_live/index.ex

```diff
@@ -390,56 +412,170 @@ defmodule ChatWeb.MainLive.Index do
+      end
+
+    {socket, uploader_data} =
+      start_chunked_upload(socket, entry, upload_key, key, secret, next_chunk)
+
     link = Helpers.upload_chunk_url(ChatWeb.Endpoint, :put, key)
 
-    {:ok, %{uploader: "UpChunk", entrypoint: link}, socket}
+    uploader_data =
+      Map.merge(%{uploader: "UpChunk", entrypoint: link, uuid: entry.uuid}, uploader_data)
+
+    {:ok, uploader_data, socket}
+  end
+
+  defp get_upload_key(%UploadEntry{} = entry, %{assigns: %{my_id: id}} = socket) do
+    destination =
+      socket
+      |> file_upload_destination()
+      |> Jason.encode!()
+      |> Base.encode64()
+
+    [
+      id,
+      destination,
+      entry.client_relative_path,
+      entry.client_name,
+      entry.client_type,
+      entry.client_size,
+      entry.client_last_modified
+    ]
+    |> Enum.join(":")
+    |> Utils.hash()
+  end
+
+  defp add_upload_to_index(socket, upload_key, key, secret) do
+    timestamp = Chat.Time.monotonic_to_unix(socket.assigns.monotonic_offset)
+    upload = %Upload{key: key, secret: secret, timestamp: timestamp}
+    UploadIndex.add(upload_key, upload)
   end
 
   def handle_chunked_progress(
-        file,
+        _name,
         %{progress: 100, uuid: uuid} = entry,
-        %{assigns: %{chunked_uploads: uploads}} = socket
+        %{assigns: %{uploads_metadata: uploads}} = socket
       ) do
     "[upload] finalizing" |> Logger.warn()
-    {key, _} = chunked_keys = uploads[uuid]
+    %UploadMetadata{} = metadata = uploads[uuid]
+    {key, _} = metadata.credentials
     ChunkedFiles.mark_consumed(key)
+    UploadIndex.delete(metadata.upload_key)
 
     "[upload] marked consumed" |> Logger.warn()
 
-    case file do
-      :dialog_file -> Page.Dialog.send_file(socket, entry, chunked_keys)
-      :room_file -> Page.Room.send_file(socket, entry, chunked_keys)
+    case metadata.destination.type do
+      :dialog -> Page.Dialog.send_file(socket, entry, metadata)
+      :room -> Page.Room.send_file(socket, entry, metadata)
     end
 
     "[upload] message sent" |> Logger.warn()
 
     socket
-    |> assign(:chunked_uploads, uploads |> Map.drop([uuid]))
+    |> assign(:uploads_metadata, Map.delete(uploads, uuid))
+    |> maybe_resume_next_upload()
     |> noreply()
     |> tap(fn _ -> "[upload] done" |> Logger.warn() end)
   end
 
-  def handle_chunked_progress(_file, _entry, socket), do: noreply(socket)
+  def handle_chunked_progress(_name, _entry, socket), do: noreply(socket)
+
+  defp start_chunked_upload(socket, entry, upload_key, key, secret, next_chunk) do
+    uploads = Map.get(socket.assigns, :uploads_metadata, %{})
+
+    active_uploads =
+      uploads
+      |> Enum.filter(fn {_uuid, metadata} -> metadata.status == :active end)
+      |> length()
+
+    status =
+      if active_uploads < @max_concurrent_uploads do
+        :active
+      else
+        :paused
+      end
+
+    metadata =
+      %UploadMetadata{}
+      |> Map.put(:credentials, {key, secret})
```

Comment:
How do we handle `secret`? 
Is it stored in DB or transferred to UI?

Reply:
As before, the secret is used for creating files in the DB, but we don't transfer it to the client side.
Therefore, we save it in the upload index for future use.